src/coreclr/jit/llvm.cpp

```diff
@@ -887,6 +925,16 @@ Value* Llvm::consumeValue(GenTree* node, Type* targetLlvmType)
                 finalValue = varTypeIsSigned(node) ? _builder.CreateSExt(nodeValue, targetLlvmType)
                                                    : _builder.CreateZExt(nodeValue, targetLlvmType);
             }
+            else if (node->OperIs(GT_LCL_VAR))
+            {
+                // implicit upcast i8 -> i32
+                assert(nodeValue->getType() == Type::getInt8Ty(_llvmContext));
+
+                GenTreeLclVarCommon* lclNode = node->AsLclVarCommon();
+                LclVarDsc* varDsc = _compiler->lvaGetDesc(lclNode->GetLclNum());
+                finalValue = varTypeIsSigned(varDsc->TypeGet()) ? _builder.CreateSExt(nodeValue, targetLlvmType)
+                                                   : _builder.CreateZExt(nodeValue, targetLlvmType);
```

Comment:
```suggestion
                // In IR, small locals are usually typed as INTs, while in LLVM registers we store them
                // with their "true" type. So here we must re-extend them (as the user requests we do so).
                LclVarDsc* varDsc = _compiler->lvaGetDesc(node->AsLclVarCommon());
                assert(varTypeIsSmall(varDsc));
                
                finalValue = varTypeIsSigned(varDsc) ? _builder.CreateSExt(nodeValue, targetLlvmType)
                                                     : _builder.CreateZExt(nodeValue, targetLlvmType);
```
Any small type can be seen here (and any small type is legal).

Reply:
Thanks!